tests/unit/coroutines_test.cc

```diff
@@ -811,3 +814,130 @@ SEASTAR_TEST_CASE(test_lambda_value_capture_continuation) {
 
 
 #endif
+
+class test_exception : std::exception { };
+
+future<> return_ex_void() {
+    fmt::print("return_ex_void()\n");
+    return make_exception_future<>(test_exception{});
+}
+
+future<> return_void() {
+    fmt::print("return_void()\n");
+    co_await sleep(1ms);
+}
+
+future<int> return_ex_int() {
+    fmt::print("return_ex_int()\n");
+    return make_exception_future<int>(test_exception{});
+}
+
+future<int> return_int() {
+    fmt::print("return_int()\n");
+    co_await sleep(1ms);
+    co_return 128;
+}
+
+template <typename T>
+struct result_wrapper {
+    T value;
+    explicit result_wrapper(T v) : value(v) {}
+};
+
+template <>
+struct result_wrapper<seastar::internal::monostate> {
+};
+
+template <bool CheckPreempt, std::invocable<> F>
+// Use result_wrapper to create a mismatch between the return type of
+// the coroutine and that of the underlying function, to ensure that
+// try_future handles this case correctly.
+future<result_wrapper<typename std::invoke_result_t<F>::value_type>>
+do_run_try_future_test(F underlying_func, int& ctor_dtor_counter, bool& run_past) {
+    const auto check_cxx_exceptions_on_exit = seastar::defer([cxx_exception_before = seastar::engine().cxx_exceptions()] () noexcept {
+        if (seastar::engine().cxx_exceptions() != cxx_exception_before) {
+            // We are in a destructor, cannot throw
+            std::abort();
+        }
+    });
```

Comment:
You could convert it to a BOOST_REQUIRE_EQUAL with a `with_cxx_exceptions_count_check(std::function<std::future<> ()>)` thing.